README.md
# TimeXt-Dart

[![Version](https://img.shields.io/badge/version-v0.3.0-blue.svg)](https://github.com/TimeXt/TimeXt-Dart/releases/tag/0.2.0)
[![Build](https://img.shields.io/badge/build-success-green.svg)](timext)
[![CodeCoverage](https://img.shields.io/badge/codeCoverage-95-green.svg)](timext)

[![License: MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/licenses/MIT)
[![Paypal](https://img.shields.io/badge/paypal-donate-blue.svg)](https://www.paypal.me/GuepardoApps)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)](http://makeapullrequest.com)

First of all many thanks to [Kizitonwose](https://github.com/kizitonwose/Time) for the original idea and already awesome library!

This library shall help to reduce code like...

```dart
int dayInMillis = 24 * 60 * 60 * 1000       // Represent a day in milliSeconds

```

## How to use

### Basics

```dart
// Type is TimeXt
var threeYears = new TimeXt(3, TimeXtUnit.Year)
var oneWeek = new TimeXt(1, TimeXtUnit.Week)
var threeDays = new TimeXt(3, TimeXtUnit.Day)
var elevenHours = new TimeXt(11, TimeXtUnit.Hour)
var sixMinutes = new TimeXt(6, TimeXtUnit.Minute)
var fiftySeconds = new TimeXt(50, TimeXtUnit.Second)
var hundredMillis = new TimeXt(100, TimeXtUnit.Millis)

var oneDayInMillis = new TimeXt(1, TimeXtUnit.Day).inMillis()  // Returns one day in milliseconds
var twoWeeksInHours = new TimeXt(2, TimeXtUnit.Week).inHours()       // Returns two weeks in hours

// Converts the existing 3h-Class into a 180min-Class
var threeHoursToMinutes = new TimeXt(3, TimeXtUnit.Hour).toMinutes()

var duration = new TimeXt(3, TimeXtUnit.Year) + new TimeXt(1, TimeXtUnit.Week) + TimeXt(50, TimeXtUnit.Second)
var multipliedDuration = 1.5 * duration
var dividedDuration = duration / 2

var isLessTrue = new TimeXt(30, TimeXtUnit.Second) < new TimeXt(30001, TimeXtUnit.Millis)
var isLessFalse = new TimeXt(3, TimeXtUnit.Week) < new TimeXt(20, TimeXtUnit.Day)

var isBiggerTrue = new TimeXt(1, TimeXtUnit.Day) > new TimeXt(23, TimeXtUnit.Hour)
var isBiggerFalse = new TimeXt(5, TimeXtUnit.Hour) > new TimeXt(301, TimeXtUnit.Minute)
```

Since version 0.2.0 TimeXt has new extensions for the long type to display this number value in a human readable string format.
Since version 0.3.0 TimeXt supports double numbers.

```dart
var readableStringFromMillis = NumberToString.formatMillis(34325055574) // 56 weeks, 5 days, 6 hours, 44 minutes, 15 seconds, 574 milliseconds
var readableStringFromSeconds = NumberToString.formatSeconds(4350554) // 7 weeks, 1 day, 8 hours, 29 minutes, 14 seconds
var readableStringFromMinutes = NumberToString.formatMinutes(432555) // 42 weeks, 6 days, 9 hours, 15 minutes
var readableStringFromHours = NumberToString.formatHours(4574) // 27 weeks, 1 day, 14 hours
var readableStringFromDays = NumberToString.formatDays(15.5) // 2 weeks, 1 day, 12 hours
```

## License

TimeXt-Dart is distributed under the MIT license. [See LICENSE](LICENSE.md) for details.

```
MIT License

Copyright (c) 2019 GuepardoApps (Jonas Schubert)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
```

## Note

This is just a prove of conecpt. Dart already has a good class with:

```dart
Duration(minutes: 1)

```

I hope Dart will suport extensions like in kotlin or C# to provide the  possibility to write things like:

```dart
var threeYears = 3.Years()

```


CHANGELOG.md
## [0.0.1] - 23.04.2019

* Initial source code

## [0.3.0] - 14.07.2019

* adds method formatDay
* refactors milliSeconds to millis
* reduces dictionary handling for smaller code
* supports doubles


test/timext_dart_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:timext_dart/timext_dart.dart';

void main() {
  test('All comparisons should work as expected', () {
    expect(new TimeXt(2, TimeXtUnit.Year), new TimeXt(730, TimeXtUnit.Day));
    expect(new TimeXt(3, TimeXtUnit.Year) > new TimeXt(20, TimeXtUnit.Day), true);
    expect(new TimeXt(0.5, TimeXtUnit.Year) < new TimeXt(200, TimeXtUnit.Day), true);

    expect(new TimeXt(2, TimeXtUnit.Week), new TimeXt(14, TimeXtUnit.Day));
    expect(new TimeXt(3, TimeXtUnit.Week) > new TimeXt(20, TimeXtUnit.Day), true);
    expect(new TimeXt(4, TimeXtUnit.Week) < new TimeXt(29, TimeXtUnit.Day), true);

    expect(new TimeXt(2, TimeXtUnit.Day), new TimeXt(48, TimeXtUnit.Hour));
    expect(new TimeXt(1, TimeXtUnit.Day) > new TimeXt(23, TimeXtUnit.Hour), true);
    expect(new TimeXt(0, TimeXtUnit.Day) < new TimeXt(1, TimeXtUnit.Hour), true);

    expect(new TimeXt(2, TimeXtUnit.Hour), new TimeXt(120, TimeXtUnit.Minute));
    expect(new TimeXt(3, TimeXtUnit.Hour) > new TimeXt(179, TimeXtUnit.Minute), true);
    expect(new TimeXt(5, TimeXtUnit.Hour) < new TimeXt(301, TimeXtUnit.Minute), true);

    expect(new TimeXt(15, TimeXtUnit.Minute), new TimeXt(900, TimeXtUnit.Second));
    expect(new TimeXt(30, TimeXtUnit.Minute) > new TimeXt(1799, TimeXtUnit.Second), true);
    expect(new TimeXt(45, TimeXtUnit.Minute) < new TimeXt(2701, TimeXtUnit.Second), true);

    expect(new TimeXt(10, TimeXtUnit.Second), new TimeXt(10000, TimeXtUnit.Millis));
    expect(new TimeXt(20, TimeXtUnit.Second) > new TimeXt(19999, TimeXtUnit.Millis), true);
    expect(new TimeXt(30, TimeXtUnit.Second) < new TimeXt(30001, TimeXtUnit.Millis), true);
  });

  test('All conversions should work as expected', () {
    expect(new TimeXt(1, TimeXtUnit.Year).inDays(), 365);
    expect(new TimeXt(14, TimeXtUnit.Day).inWeeks(), 2);
    expect(new TimeXt(17520, TimeXtUnit.Hour).inYears(), 2);
    expect(new TimeXt(5.5, TimeXtUnit.Minute).inSeconds(), 330);
    expect(new TimeXt(1.52, TimeXtUnit.Second).inMillis(), 1520);
    expect(new TimeXt(30000, TimeXtUnit.Millis).inMinutes(), 0.5);
  });

  test('All casts should work as expected', () {
    expect(new TimeXt(1, TimeXtUnit.Year).toDays(), new TimeXt(365, TimeXtUnit.Day));
    expect(new TimeXt(14, TimeXtUnit.Day).toWeeks(), new TimeXt(2, TimeXtUnit.Week));
    expect(new TimeXt(17520, TimeXtUnit.Hour).toYears(), new TimeXt(2, TimeXtUnit.Year));
    expect(new TimeXt(5.5, TimeXtUnit.Minute).toSeconds(), new TimeXt(330, TimeXtUnit.Second));
    expect(new TimeXt(1.52, TimeXtUnit.Second).toMillis(), new TimeXt(1520, TimeXtUnit.Millis));
    expect(new TimeXt(30000, TimeXtUnit.Millis).toMinutes(), new TimeXt(0.5, TimeXtUnit.Minute));
  });

  test('Duration casts should work as expected', () {
    expect(TimeXt.fromDuration(Duration(minutes: 1), unit: TimeXtUnit.Second), new TimeXt(60, TimeXtUnit.Second));
    expect(new TimeXt(60, TimeXtUnit.Second).toDuration(), Duration(milliseconds: 60000));
  });

  test('All NumberToString methods should work as expected', () {
    expect(NumberToString.formatMillis(0), "0 milliseconds");
    expect(NumberToString.formatSeconds(0), "0 milliseconds");
    expect(NumberToString.formatMinutes(0), "0 milliseconds");
    expect(NumberToString.formatHours(0), "0 milliseconds");
    expect(NumberToString.formatDays(0), "0 milliseconds");

    expect(NumberToString.formatMillis(34325055574), "56 weeks, 5 days, 6 hours, 44 minutes, 15 seconds, 574 milliseconds");
    expect(NumberToString.formatSeconds(4350554), "7 weeks, 1 day, 8 hours, 29 minutes, 14 seconds");
    expect(NumberToString.formatSeconds(90.250), "1 minute, 30 seconds, 250 milliseconds");
    expect(NumberToString.formatMinutes(432555), "42 weeks, 6 days, 9 hours, 15 minutes");
    expect(NumberToString.formatMinutes(0.5), "30 seconds");
    expect(NumberToString.formatHours(4574), "27 weeks, 1 day, 14 hours");
    expect(NumberToString.formatHours(49.25), "2 days, 1 hour, 15 minutes");
    expect(NumberToString.formatDays(23), "3 weeks, 2 days");
    expect(NumberToString.formatDays(15.5), "2 weeks, 1 day, 12 hours");
  });
}


lib/timext_dart.dart
library timext_dart;

class TimeXt {
  final double _value;
  final TimeXtUnit _unit;

  TimeXt(this._value, this._unit);

  double inYears() => this._value * this._unit._value / TimeXtUnit.Year._value;
  double inWeeks() => this._value * this._unit._value / TimeXtUnit.Week._value;
  double inDays() => this._value * this._unit._value / TimeXtUnit.Day._value;
  double inHours() => this._value * this._unit._value / TimeXtUnit.Hour._value;
  double inMinutes() => this._value * this._unit._value / TimeXtUnit.Minute._value;
  double inSeconds() => this._value * this._unit._value / TimeXtUnit.Second._value;
  double inMillis() => this._value * this._unit._value / TimeXtUnit.Millis._value;

  TimeXt toYears() => TimeXt(this.inYears(), TimeXtUnit.Year);
  TimeXt toWeeks() => TimeXt(this.inWeeks(), TimeXtUnit.Week);
  TimeXt toDays() => TimeXt(this.inDays(), TimeXtUnit.Day);
  TimeXt toHours() => TimeXt(this.inHours(), TimeXtUnit.Hour);
  TimeXt toMinutes() => TimeXt(this.inMinutes(), TimeXtUnit.Minute);
  TimeXt toSeconds() => TimeXt(this.inSeconds(), TimeXtUnit.Second);
  TimeXt toMillis() => TimeXt(this.inMillis(), TimeXtUnit.Millis);

  static TimeXt fromDuration(Duration duration, {TimeXtUnit unit: TimeXtUnit.Millis}) => TimeXt(duration.inMilliseconds.toDouble() * TimeXtUnit.Millis._value / unit._value, unit);
  Duration toDuration() => Duration(milliseconds: this.inMillis().toInt());

  operator +(TimeXt other) => TimeXt(((this.inMillis() + other.inMillis()) / this._unit._value) * TimeXtUnit.Millis._value, this._unit);
  operator -(TimeXt other) => TimeXt(((this.inMillis() - other.inMillis()) / this._unit._value) * TimeXtUnit.Millis._value, this._unit);
  operator *(double other) => other == 0 ? throw new NullThrownError() : TimeXt(this._value * other, this._unit);
  operator /(double other) => TimeXt(this._value / other, this._unit);
  operator ==(other) => this.inMillis() == other.inMillis();
  operator >(other) => this.inMillis() > other.inMillis();
  operator <(other) => this.inMillis() < other.inMillis();

  @override int get hashCode => this.inMillis().hashCode;
  @override String toString() => "$_value, $_unit";
}

class TimeXtUnit {
  final double _value;

  const TimeXtUnit._internal(this._value);

  @override String toString() => "$_value";

  static const Year = const TimeXtUnit._internal(365 * 24 * 60 * 60 * 1e3);
  static const Week = const TimeXtUnit._internal(7 * 24 * 60 * 60 * 1e3);
  static const Day = const TimeXtUnit._internal(24 * 60 * 60 * 1e3);
  static const Hour = const TimeXtUnit._internal(60 * 60 * 1e3);
  static const Minute = const TimeXtUnit._internal(60 * 1e3);
  static const Second = const TimeXtUnit._internal(1e3);
  static const Millis = const TimeXtUnit._internal(1e0);
}

class NumberToString {
  static List _dictionaryToStringList(double time, double divider) {
    List stringList = [];
    var dictionary = {
      "week": [7 * 24 * 60 * 60 * 1e3, 2 ^ 64],
      "day": [24 * 60 * 60 * 1e3, 7],
      "hour": [60 * 60 * 1e3, 24],
      "minute": [60 * 1e3, 60],
      "second": [1e3, 60],
      "millisecond": [1, 1e3]
    };
    dictionary.forEach((key, value) {
      var testValue = ((time / (value.elementAt(0) / divider)) % value.elementAt(1)).floor();
      testValue > 1
          ? stringList.add("$testValue ${key}s")
          : testValue > 0
              ? stringList.add("$testValue $key")
              : stringList.add("");
    });
    return stringList.where((value) => value.length > 0).toList();
  }

  static String formatMillis(double millis) {
    List stringList = _dictionaryToStringList(millis, 1);
    return stringList.length > 0 ? stringList.join(", ") : "0 milliseconds";
  }

  static String formatSeconds(double seconds) {
    List stringList = _dictionaryToStringList(seconds, 1e3);
    return stringList.length > 0 ? stringList.join(", ") : formatMillis(seconds * 1e3);
  }

  static String formatMinutes(double minutes) {
    List stringList = _dictionaryToStringList(minutes, 60 * 1e3);
    return stringList.length > 0 ? stringList.join(", ") : formatSeconds(minutes * 60);
  }

  static String formatHours(double hours) {
    List stringList = _dictionaryToStringList(hours, 60 * 60 * 1e3);
    return stringList.length > 0 ? stringList.join(", ") : formatMinutes(hours * 60);
  }

  static String formatDays(double days) {
    List stringList = _dictionaryToStringList(days, 24 * 60 * 60 * 1e3);
    return stringList.length > 0 ? stringList.join(", ") : formatMinutes(days * 24);
  }
}
